File: 0079-word-search/0079-word-search.java
class Solution {
    int[][] directions = {{1,0}, {0,1}, {-1,0}, {0,-1}};
    boolean dfs(char[][] board, String word, int i, int j, int s, boolean[][] vis) {
        if(s == word.length()-1) return true;
        vis[i][j] = true;
        for(int[] dir:directions) {
            int x = dir[0] + i;
            int y = dir[1] + j;
            if(x < 0 || y < 0 || x >= board.length || y >= board[0].length || vis[x][y] || (s+1 < word.length() && board[x][y] != word.charAt(s+1))) continue;
            else {
                if(dfs(board, word, x, y, s+1, vis)) return true;
            }
        }
        vis[i][j] = false;
        return false;
    }
    public boolean exist(char[][] board, String word) {
        boolean[][] vis = new boolean[board.length][board[0].length];
        for(int i=0; i<board.length; i++) {
            for(int j=0; j<board[0].length; j++) {
                if(board[i][j] == word.charAt(0)) {
                    if(dfs(board, word, i, j, 0, vis)) return true;
                }
            }
        }
        return false;
    }
}

File: 0079-word-search/0079-word-search.py
class Solution:
    directions = [[1,0], [0,1], [-1,0], [0,-1]]
    def dfs(self, board: List[List[str]], word: str, i: int, j: int, s: int, vis: List[List[bool]]) -> bool:
        if s == len(word) - 1 : return True
        vis[i][j] = True
        for dir in self.directions :
            x = i + dir[0]
            y = j + dir[1]
            if x < 0 or y < 0 or x >= len(board) or y >= len(board[0]) or board[x][y] != word[s+1] or vis[x][y] : continue
            else :
                if self.dfs(board, word, x, y, s+1, vis) : return True
        vis[i][j] = False
        return False
    def exist(self, board: List[List[str]], word: str) -> bool:
        if len(word) > (len(board) * len(board[0])) : return False
        vis = [[False for j in range(len(board[0]))] for i in range(len(board))]
        for i in range(len(board)) :
            for j in range(len(board[0])) :
                if board[i][j] == word[0] :
                    if self.dfs(board, word, i, j, 0, vis) : return True
        return False